### ISSUE TITLE
Export clean ESM modules

### ISSUE THREAD

**Issue by sebdeckers** [graphql/graphql-js#1819]
Using graphql@14.2.1

`/node_modules/graphql/jsutils/instanceOf.mjs` line 16

```js
// See: https://expressjs.com/en/advanced/best-practice-performance.html#set-node_env-to-production
// See: https://webpack.js.org/guides/production/
export default process.env.NODE_ENV === 'production' ? // eslint-disable-next-line no-shadow
```

```
Uncaught ReferenceError: process is not defined
    at instanceOf.mjs:16
```

The check `process.env.NODE_ENV === 'production'` breaks in the browser when using ESM version of `graphql` package. Is it possible to remove before publishing to NPM? Without it the ESM files work perfectly.

I am not using Webpack/RollUp/Babel. Simply copying the ESM files from node_modules to my frontend folder using [unbundle](https://www.npmjs.com/package/unbundle).

Related but different:
- https://github.com/graphql/graphql-js/issues/1536
- https://github.com/graphql/graphql-js/issues/1293
- https://github.com/graphql/graphql-js/issues/1217

**Comment by abdonrd** [comment]
Also related but different: https://github.com/apollographql/graphql-tag/issues/229

**Comment by Carglass** [comment]
Hello, 

Id like to try to help with this issue. Using Unbundle, do you have a possibility to differentiate production from development code? I could not find a way to do it without using "bundlers". If there is not, the main issue, is that this export returns an optimized function for production, so that in the absence of the process.env, there is a choice to make about using the optimized function or the more helpful for development function. Any thoughts on this?

**Comment by dandv** [comment]
Not sure if this will get any traction before Node 12 drops the --experimental-modules flag, because they've asked the community to [not publish any ES modules until then](https://twitter.com/sheerun/status/1122894214646370304).

**Comment by sebdeckers** [comment]
> While we are aware that the community has embraced the `"module"` field, it is unlikely that Node.js will adopt that field as many of the packages published using `"module"` include ES module JavaScript that may not evaluate in Node.js (because extensions are left off of filenames, or the code includes `require` statements, etc.). Please do not publish any ES module packages intended for use by Node.js until this is resolved.

1. I am looking to run the ESM on the web, not Node.js. (Universal ESM would be nice tho.)
1. The `module` field (et al.) is a de-facto practice. Supported by WebPack, RollUp, and Pika (afaik).

Compromise: Publish a distinct ESM package on NPM. Avoids conflict with Node.js and is backwards compatible with a variety of bundlers. E.g. [lodash-es](https://www.npmjs.com/package/lodash-es)

Caveat, I'm not familiar with the reasons behind the above statement. Maybe this idea would cause problems in the long run. 🤷🏼‍♂️

**Comment by abdonrd** [comment]
Any news here? Thanks in advance!

**Comment by Banou26** [comment]
Could the ESM exports use complete import paths ?
e.g
```diff
- export { Source } from './source';
+ export { Source } from './source.mjs';
```

Since this results in errors like
```
Error: Cannot find module C:\dev\oz\packages\api\node_modules\graphql\language\source imported from C:\dev\oz\packages\api\node_modules\graphql\language\index.mjs
```
when using nodejs with ESM

**Comment by jaydenseric** [comment]
@Banou26 to achieve that, [`babel-plugin-transform-require-extensions`](https://github.com/jaydenseric/babel-plugin-transform-require-extensions) can be used.

For example, here it is in action with `extract-files`:

- The source is propper Node.js compatible ESM, with imports using `.mjs` extensions: https://github.com/jaydenseric/extract-files/blob/v7.0.0/src/index.mjs#L1
- Babel is configured to replace the `.mjs` extensions with `.js` when the target format is CJS: https://github.com/jaydenseric/extract-files/blob/v7.0.0/babel.config.js#L7

**Comment by arlac77** [comment]
@jaydenseric what is i want to use the raw npm content without further bundling either from node or the browser than the content must be 100% valid esm with full file names.

**Comment by jaydenseric** [comment]
@arlac77 it is:

https://unpkg.com/extract-files@7.0.0/lib/index.mjs

**Comment by arlac77** [comment]
> @arlac77 it is:
> 
> https://unpkg.com/extract-files@7.0.0/lib/index.mjs

but deep inside there are still some imports without extension
```sh
Error: Cannot find module /Users/markus/build/hook-ci/node_modules/graphql/jsutils/devAssert imported from /Users/markus/build/hook-ci/node_modules/graphql/utilities/buildASTSchema.mjs
```

**Comment by arlac77** [comment]
> @arlac77 it is:
> 
> https://unpkg.com/extract-files@7.0.0/lib/index.mjs

Thanks @jaydenseric,
submitted PR [2451](https://github.com/graphql/graphql-js/pull/2451)

**Comment by IvanGoncharov** [comment]
Released as `15.0.0-rc.2`.
Last remaining issue tracked here: #2409
